### PR TITLE
samples: net: coap_server: Fixes for retransmission and observer

### LINKED ISSUE #38697: The coap_server sample is missing the actual send in the retransmit routine
**Describe the bug**
If a device can't reach it's destination, it will get into `retransmit_request` for several times to decrement its retries, however no actual send is performed which makes the retransmit request useless.

**To Reproduce**
1. Build the coap_server sample
2. Observe the resource i.e. `coap get -o coap://xx.xx.xx.xx/obs`
3. Stop the coap client
4. Wait until the coap server gets into the retransmit_request

**Expected behavior**
I would expect the coap server to actual resend a message instead of just decrementing the retries.

**Impact**
It leaves the coap server retry mechanism useless.

**Environment (please complete the following information):**
 - Zephyr v2.6.0

### LINKED ISSUE #38700: Observable resource of coap server seems to not support 2 observers simultaneously
**Describe the bug**
If I try to observe the `/obs` resource from my pc, I do get the desired behaviour of the counter being incremented and transmitted every 5 seconds. If I however open a second terminal and observe the `/obs` resource for a second time as well, the first observer keeps receiving counter updates however my second observer doesn't. It received the counter value only a single time. 

**To Reproduce**
1. Build the coap_server sample
2. Observe the resource i.e. coap get -o coap://xx.xx.xx.xx/obs
3. Open another terminal and observe the resource again coap get -o coap://xx.xx.xx.xx/obs
4. Only first observer keeps receiving counter updates
5. The second observer will only receive a single counter update

**Expected behavior**
Both observers must receive the same counter updates.

**Impact**
I am not able to have 2 observers observing the same resource.

**Logs and console output**
Observer 0

```
coap get -o coap://xx.xx.xx.xx/obs
(2.05)  Counter: 0

(2.05)  Counter: 1

(2.05)  Counter: 2

```

Observer 1

```
coap get -o coap://xx.xx.xx.xx/obs
(2.05)  Counter: 0
```

Coap server

```
*** Booting Zephyr OS build zephyr-v2.6.0  ***


[00:00:00.057,000] <inf> net_config: Initializing network
[00:00:00.057,000] <inf> net_config: Waiting interface 1 (0x80000dc8) to be up...
[00:00:03.011,000] <inf> net_config: Interface 1 (0x80000dc8) coming up
[00:00:03.011,000] <inf> eth_mcux: ETH_0 enabled 100M full-duplex mode.
[00:00:03.011,000] <inf> net_config: Running dhcpv4 client...
[00:00:12.066,000] <inf> net_dhcpv4: Received: xx.xx.xx.xx
[00:00:12.066,000] <inf> net_config: IPv4 address: xx.xx.xx.xx
[00:00:12.066,000] <inf> net_config: Lease time: 604800 seconds
[00:00:12.066,000] <inf> net_config: Subnet: xx.xx.xx.xx
[00:00:12.066,000] <inf> net_config: Router: xx.xx.xx.xx
[00:00:12.066,000] <dbg> net_coap_server_sample.main: Start CoAP-server sample
[00:00:14.349,000] <inf> net_coap_server_sample: *******
[00:00:14.350,000] <inf> net_coap_server_sample: type: 0 code 1 id 18191
[00:00:14.350,000] <inf> net_coap_server_sample: *******
[00:00:14.350,000] <dbg> net_coap_server_sample: Response
                                                 64 45 47 0f 98 f9 a6 54  61 02 60 ff 43 6f 75 6e |dEG....T a.`.Coun
                                                 74 65 72 3a 20 30 0a                             |ter: 0.
[00:00:16.694,000] <inf> net_coap_server_sample: *******
[00:00:16.694,000] <inf> net_coap_server_sample: type: 0 code 1 id 9353
[00:00:16.694,000] <inf> net_coap_server_sample: *******
[00:00:16.694,000] <dbg> net_coap_server_sample: Response
                                                 64 45 24 89 6c 20 c4 f0  61 02 60 ff 43 6f 75 6e |dE$.l .. a.`.Coun
                                                 74 65 72 3a 20 30 0a                             |ter: 0.
[00:00:21.694,000] <dbg> net_coap_server_sample: Response
                                                 44 45 43 8a 98 f9 a6 54  61 03 60 ff 43 6f 75 6e |DEC....T a.`.Coun
                                                 74 65 72 3a 20 31 0a                             |ter: 1.
[00:00:21.748,000] <dbg> net_coap_server_sample: Response
                                                 44 45 43 8b 6c 20 c4 f0  61 03 60 ff 43 6f 75 6e |DEC.l .. a.`.Coun
                                                 74 65 72 3a 20 31 0a                             |ter: 1.
[00:00:24.070,000] <inf> net_coap_server_sample: RETRANSMIT (17291)(4)
[00:00:26.748,000] <dbg> net_coap_server_sample: Response
                                                 44 45 43 8c 98 f9 a6 54  61 04 60 ff 43 6f 75 6e |DEC....T a.`.Coun
                                                 74 65 72 3a 20 32 0a                             |ter: 2.
[00:00:26.802,000] <dbg> net_coap_server_sample: Response
                                                 44 45 43 8d 6c 20 c4 f0  61 04 60 ff 43 6f 75 6e |DEC.l .. a.`.Coun
                                                 74 65 72 3a 20 32 0a                             |ter: 2.


```

**Environment (please complete the following information):**
 - Observers: Windows
 - Zephyr V2.6.0

### LINKED ISSUE #38701: Observable resource of coap server seems to not support a restart of an observer
**Describe the bug**
If I start an observer for the `/obs` resource, and stop it after some time and then restart an observer, the counter will be received only a single time and will be never received again in the future.

**To Reproduce**
1. Build the coap_server sample
2. Observe the resource i.e. coap get -o coap://xx.xx.xx.xx/obs
3. Abruptly stop the observer 
4. Observe the resource i.e. coap get -o coap://xx.xx.xx.xx/obs
5. Wait and see that no counter updates are being received anymore

**Expected behavior**
I would expect that the coap server sending counter updates to my second observer and also performs retransmits to my first observer.

**Impact**
I am not able to stop an observer and restart it afterwards.

**Logs and console output**

*** Booting Zephyr OS build zephyr-v2.6.0  ***
Observer

```
coap get -o coap://xx.xx.xx.xx/obs

(2.05)  Counter: 0

(2.05)  Counter: 1

(2.05)  Counter: 2

^CTerminate batch job (Y/N)?

coap get -o coap://xx.xx.xx.xx/obs

(2.05)  Counter: 3

```

Coap Server
```
[00:00:00.057,000] <inf> net_config: Initializing network
[00:00:00.057,000] <inf> net_config: Waiting interface 1 (0x80000dc8) to be up...
[00:00:03.011,000] <inf> net_config: Interface 1 (0x80000dc8) coming up
[00:00:03.011,000] <inf> eth_mcux: ETH_0 enabled 100M full-duplex mode.
[00:00:03.011,000] <inf> net_config: Running dhcpv4 client...
[00:00:04.374,000] <err> net_arp: Gateway not set for iface 0x80000dc8
[00:00:04.481,000] <err> net_arp: Gateway not set for iface 0x80000dc8
[00:00:05.027,000] <inf> net_dhcpv4: Received: xx.xx.xx.xx
[00:00:05.027,000] <inf> net_config: IPv4 address: xx.xx.xx.xx
[00:00:05.027,000] <inf> net_config: Lease time: 604800 seconds
[00:00:05.027,000] <inf> net_config: Subnet: xx.xx.xx.xx
[00:00:05.027,000] <inf> net_config: Router: xx.xx.xx.xx
[00:00:05.027,000] <dbg> net_coap_server_sample.main: Start CoAP-server sample
[00:00:13.209,000] <inf> net_coap_server_sample: *******
[00:00:13.209,000] <inf> net_coap_server_sample: type: 0 code 1 id 24601
[00:00:13.209,000] <inf> net_coap_server_sample: *******
[00:00:13.209,000] <dbg> net_coap_server_sample: Response
                                                 64 45 60 19 f4 3d ad ac  61 02 60 ff 43 6f 75 6e |dE`..=.. a.`.Coun
                                                 74 65 72 3a 20 30 0a                             |ter: 0.
[00:00:18.210,000] <dbg> net_coap_server_sample: Response
                                                 44 45 d1 e6 f4 3d ad ac  61 03 60 ff 43 6f 75 6e |DE...=.. a.`.Coun
                                                 74 65 72 3a 20 31 0a                             |ter: 1.
[00:00:23.210,000] <dbg> net_coap_server_sample: Response
                                                 44 45 d1 e7 f4 3d ad ac  61 04 60 ff 43 6f 75 6e |DE...=.. a.`.Coun
                                                 74 65 72 3a 20 32 0a                             |ter: 2.
[00:00:28.211,000] <dbg> net_coap_server_sample: Response
                                                 44 45 d1 e8 f4 3d ad ac  61 05 60 ff 43 6f 75 6e |DE...=.. a.`.Coun
                                                 74 65 72 3a 20 33 0a                             |ter: 3.
[00:00:33.196,000] <inf> net_coap_server_sample: *******
[00:00:33.196,000] <inf> net_coap_server_sample: type: 0 code 1 id 59996
[00:00:33.196,000] <inf> net_coap_server_sample: *******
[00:00:33.196,000] <dbg> net_coap_server_sample: Response
                                                 64 45 ea 5c 2d 6d 19 a6  61 05 60 ff 43 6f 75 6e |dE.\-m.. a.`.Coun
                                                 74 65 72 3a 20 33 0a                             |ter: 3.
[00:00:38.196,000] <dbg> net_coap_server_sample: Response
                                                 44 45 d1 e9 f4 3d ad ac  61 06 60 ff 43 6f 75 6e |DE...=.. a.`.Coun
                                                 74 65 72 3a 20 34 0a                             |ter: 4.
[00:00:38.249,000] <dbg> net_coap_server_sample: Response
                                                 44 45 d1 ea 2d 6d 19 a6  61 06 60 ff 43 6f 75 6e |DE..-m.. a.`.Coun
                                                 74 65 72 3a 20 34 0a                             |ter: 4.

```

**Environment (please complete the following information):**
 - Observer: Windows
 - Zephyr V2.6.0

### LINKED ISSUE #38702: Coap server not properly removing observers
**Describe the bug**
If a host is observing a resource and then abruptly stops observing it, I would expect the coap server to do multiple retransmits (up till now the actual send is missing), and then remove the observer. Now only a coap message with type `COAP_TYPE_RESET` will remove an observer. Besides that, the free spaces in the buffer of observers runs empty after 3 times of starting up an observer and properly removing it since the observer in the observers array needs a `memset(o, 0, sizeof(struct coap_observer));` after removal to be able to reuse that observer spot in the buffer again in the future.

**To Reproduce**
1. Build the coap_server sample
2. Observe the resource i.e. `coap get -o coap://xx.xx.xx.xx/obs`
3. Abruptly stop the observer
4. Observe the resource i.e. `coap get -o coap://xx.xx.xx.xx/obs`
5. Abruptly stop the observer
6. Observe the resource i.e. `coap get -o coap://xx.xx.xx.xx/obs`
7. Abruptly stop the observer
8. Observe the resource i.e. `coap get -o coap://xx.xx.xx.xx/obs`
9. See that no counter value is received anymore and the warning in the console of the coap server: `<wrn> net_coap_server_sample: No handler for such request (-12)`

**Expected behavior**
I expect the observer to be cleaned up properly if:
* it receives a message with type `COAP_TYPE_RESET`
* 4 retransmits fail to the observer

Now only the first is implemented and even for that case, the observer is not cleaned up properly and the spot in the observers array will never be able to be reused since `coap_observer_next_unused` will still recognize that spot as used. A `memset(o, 0, sizeof(struct coap_observer));` of that particular observer is necessary to free the space of the observer that is removed.

**Impact**
The coap server will try to notify an observer that is not longer observing a resource.
Even when 3 observers are removed, the spots in the observers array won't be free for a new observer since those observers are not reset in the observers array.

**Logs and console output**
Observer

```
coap get -o coap://172.149.20.44/obs
(2.05)  Counter: 0
^CTerminate batch job (Y/N)?
coap get -o coap://172.149.20.44/obs
(2.05)  Counter: 0
^CTerminate batch job (Y/N)?
coap get -o coap://172.149.20.44/obs
(2.05)  Counter: 0
^CTerminate batch job (Y/N)?
coap get -o coap://172.149.20.44/obs

```

Coap server
```
*** Booting Zephyr OS build zephyr-v2.6.0  ***


[00:00:00.057,000] <inf> net_config: Initializing network
[00:00:00.057,000] <inf> net_config: Waiting interface 1 (0x80000dc8) to be up..                                                                                                                                                             .
[00:00:03.011,000] <inf> net_config: Interface 1 (0x80000dc8) coming up
[00:00:03.011,000] <inf> eth_mcux: ETH_0 enabled 100M full-duplex mode.
[00:00:03.011,000] <inf> net_config: Running dhcpv4 client...
[00:00:05.097,000] <err> net_arp: Gateway not set for iface 0x80000dc8
[00:00:06.026,000] <inf> net_dhcpv4: Received: xx.xx.xx.xx
[00:00:06.026,000] <inf> net_config: IPv4 address: xx.xx.xx.xx
[00:00:06.026,000] <inf> net_config: Lease time: 604800 seconds
[00:00:06.026,000] <inf> net_config: Subnet: xx.xx.xx.xx
[00:00:06.026,000] <inf> net_config: Router: xx.xx.xx.xx
[00:00:06.026,000] <dbg> net_coap_server_sample.main: Start CoAP-server sample
[00:00:11.594,000] <inf> net_coap_server_sample: *******
[00:00:11.594,000] <inf> net_coap_server_sample: type: 0 code 1 id 3151
[00:00:11.594,000] <inf> net_coap_server_sample: *******
[00:00:11.594,000] <dbg> net_coap_server_sample: Response
                                                 64 45 0c 4f bf 6e 7c 9e  61 02                                                                                                                                                              60 ff 43 6f 75 6e |dE.O.n|. a.`.Coun
                                                 74 65 72 3a 20 30 0a                                                                                                                                                                                          |ter: 0.
[00:00:13.997,000] <inf> net_coap_server_sample: *******
[00:00:13.997,000] <inf> net_coap_server_sample: type: 0 code 1 id 30119
[00:00:13.997,000] <inf> net_coap_server_sample: *******
[00:00:13.997,000] <dbg> net_coap_server_sample: Response
                                                 64 45 75 a7 11 d2 8e a0  61 02                                                                                                                                                              60 ff 43 6f 75 6e |dEu..... a.`.Coun
                                                 74 65 72 3a 20 30 0a                                                                                                                                                                                          |ter: 0.
[00:00:16.887,000] <inf> net_coap_server_sample: *******
[00:00:16.887,000] <inf> net_coap_server_sample: type: 0 code 1 id 60065
[00:00:16.887,000] <inf> net_coap_server_sample: *******
[00:00:16.887,000] <dbg> net_coap_server_sample: Response
                                                 64 45 ea a1 05 e7 9e d1  61 02                                                                                                                                                              60 ff 43 6f 75 6e |dE...... a.`.Coun
                                                 74 65 72 3a 20 30 0a                                                                                                                                                                                          |ter: 0.
[00:00:20.064,000] <wrn> net_coap_server_sample: No handler for such request (-1                                                                                                                                                             2)
```

**Environment (please complete the following information):**
 - Observer: Windows
 - Zephyr V2.6.0

### PR DESCRIPTION
**Pull request by rlubos**
Several fixes for the sample for better user experience:
 * Fix retransmission mechanism, to actually send the retransmitted message and fix the retransmission timeout calculation.
 * Remove observer on notification timeout to release resources.
 * Finetune some of the sample parameters, to improve sample responsiveness.

Fixes #38697
Fixes #38700
Fixes #38701
Fixes #38702

### CODE REVIEW

**Comment by rlubos** [comment]
CC @AllemanMa - The sample should handle observations better now, please have a look.

**Comment by carlocaione** [comment]
`Fixes #38734` is this correct?

**Comment by rlubos** [comment]
> `Fixes #38734` is this correct?

Apologies, obviously not. Not sure where did I get that from 🤔

**Review by cfriedt — APPROVED** [review]

**Comment by AllemanMa** [comment]
@rlubos, I tested the sample and observations seem to work fine now. Thank you! When I monitor the console, I notice regularly retransmissions. Is that because the coap server is trying to send to my 3 observers simultaneously instead of scheduling it?

**Comment by rlubos** [comment]
> When I monitor the console, I notice regularly retransmissions. Is that because the coap server is trying to send to my 3 observers simultaneously instead of scheduling it?

I'm not sure what do you mean with the last sentence. Every 5 seconds, the sample updates the observe counter, and tries to send an updated value to each registered observer (each notification will print it's hexdup descibed as "Response").

If you see a hexdump marked as "Retrasmit" however, this means that some of the previous notifications could not reach the observer (did you kill any?). In such case, the sample will take a few retransmission attempts after which the unreachable observer is removed from the observer list. After this event you should no longer see retransmissions (provided there is no other issue with the network of course). But if you keep turning observers on and off, the retransmissions will show up from time to time.

**Comment by AllemanMa** [comment]
When I start 3 observers that observe a resource on the coap server (without stopping or starting the observer), I can see that  sometimes 1 or 2 out of 3 observers do get its counter value after a retransmission instead of immediately. That is in fact not a problem for me, I just noticed this and was wondering whether or not it was caused by the coap server.  As you can see below, counter value 69 had to be retransmitted for 2 out of 3 observers, but for counter value 70, no retransmissions were necessary. Do you understand what I mean?

```
[00:06:11.445,000] <dbg> net_coap_server_sample: Response
                                                 44 45 23 63 21 d5 dc 01  61 47 60 ff 43 6f 75 6e |DE#c!... aG`.Coun
                                                 74 65 72 3a 20 36 39 0a                          |ter: 69.
[00:06:11.446,000] <dbg> net_coap_server_sample: Response
                                                 44 45 23 64 84 8f 36 a8  61 47 60 ff 43 6f 75 6e |DE#d..6. aG`.Coun
                                                 74 65 72 3a 20 36 39 0a                          |ter: 69.
[00:06:11.446,000] <dbg> net_coap_server_sample: Response
                                                 44 45 23 65 43 09 95 1a  61 47 60 ff 43 6f 75 6e |DE#eC... aG`.Coun
                                                 74 65 72 3a 20 36 39 0a                          |ter: 69.
[00:06:13.629,000] <dbg> net_coap_server_sample: Retrasmit
                                                 44 45 23 64 84 8f 36 a8  61 47 60 ff 43 6f 75 6e |DE#d..6. aG`.Coun
                                                 74 65 72 3a 20 36 39 0a                          |ter: 69.
[00:06:14.411,000] <dbg> net_coap_server_sample: Retrasmit
                                                 44 45 23 65 43 09 95 1a  61 47 60 ff 43 6f 75 6e |DE#eC... aG`.Coun
                                                 74 65 72 3a 20 36 39 0a                          |ter: 69.
[00:06:16.447,000] <dbg> net_coap_server_sample: Response
                                                 44 45 23 66 21 d5 dc 01  61 48 60 ff 43 6f 75 6e |DE#f!... aH`.Coun
                                                 74 65 72 3a 20 37 30 0a                          |ter: 70.
[00:06:16.447,000] <dbg> net_coap_server_sample: Response
                                                 44 45 23 67 84 8f 36 a8  61 48 60 ff 43 6f 75 6e |DE#g..6. aH`.Coun
                                                 74 65 72 3a 20 37 30 0a                          |ter: 70.
[00:06:16.447,000] <dbg> net_coap_server_sample: Response
                                                 44 45 23 68 43 09 95 1a  61 48 60 ff 43 6f 75 6e |DE#hC... aH`.Coun
                                                 74 65 72 3a 20 37 30 0a                          |ter: 70.
```

**Comment by rlubos** [comment]
@AllemanMa Thanks for clarifying, now I understand what your problem is, hovewer I observe no such issue in emulated environment (qemu). It seems that either the notification or the acknowledgment is lost somewhere in the net stack or less likely on the host side.

W/o additional info, it's hard to tell where exact the problem is. Personally I'd try increasing `NET_PKT_RX_COUNT`/`NET_PKT_TX_COUNT`/`NET_BUF_TX_COUNT`/`NET_PKT_RX_COUNT` to see if it helps - not sure which network driver do you use, but some of them might drop packets silently if they fail to allocate packet (w/o specifying allocation timeout). It'd be good to run wireshark on the host side as well, and see if the initial notification before retransmission is actually received on the interface and if the acknowledgment is sent. That should help to identify if the problem is with TX or RX side of the Zephyr net stack.

**Comment by AllemanMa** [comment]
@rlubos The situation seems to occur at TX side of the coap server when I verify with Wireshark. I tried increasing all 4 configs but neither of them seemed to help. I am using the development board MIMXRT1064-EVK.

**Comment by rlubos** [comment]
I'm not familiar with this board, nor the driver it uses. I think it's time to attach the debugger and see where the packet is lost (or perhaps there are some extra logs that could be enabled at the driver level).

File: samples/net/sockets/coap_server/src/coap-server.c [hunk 1]
```diff
@@ -945,28 +951,53 @@ static int large_create_post(struct coap_resource *resource,
...
 		return;
 	}
 
 	if (!coap_pending_cycle(pending)) {
+		remove_observer(&pending->addr);
 		k_free(pending->data);
 		coap_pending_clear(pending);
-		return;
+	} else {
+		net_hexdump("Retrasmit", pending->data, pending->len);
```
Comment (pdgendt) [inline samples/net/sockets/coap_server/src/coap-server.c:991]:
```suggestion
		net_hexdump("Retransmit", pending->data, pending->len);
```

File: samples/net/sockets/coap_server/src/coap-server.c [hunk 2]
```diff
@@ -1309,6 +1335,28 @@ static struct coap_resource *find_resouce_by_observer(
...
+{
+	struct coap_resource *r;
+	struct coap_observer *o;
+
+	o = coap_find_observer_by_addr(observers, NUM_OBSERVERS, addr);
+	if (!o) {
+		return;
+	}
+
+	r = find_resouce_by_observer(resources, o);
```
Comment (pdgendt) [inline samples/net/sockets/coap_server/src/coap-server.c:1348]: There's a typo in the function name `find_resource_by_observer`.

**Review by pdgendt — APPROVED** [review]
Some small typo nits

**Comment by rlubos** [comment]
@pdgendt Thanks, fixed both typos.

**Review by nashif — APPROVED** [review]

**Pull request merged**